cpp/oneapi/dal/backend/primitives/sort/sort_dpc.cpp

```diff
@@ -0,0 +1,530 @@
+            }
+
+            const std::uint32_t n_groups = item.get_group_range(0);
+            const std::uint32_t n_sub_groups = sbg.get_group_range()[0];
+            const std::uint32_t n_total_sub_groups = n_sub_groups * n_groups;
+            const IndexType elems_for_sbg =
+                elem_count / n_total_sub_groups + bool(elem_count % n_total_sub_groups);
+            const std::uint32_t local_size = item.get_sub_group().get_local_range()[0];
+
+            const std::uint32_t local_id = item.get_sub_group().get_local_id()[0];
+            const std::uint32_t sub_group_id = sbg.get_group_id();
+            const std::uint32_t group_id = item.get_group().get_id(0) * n_sub_groups + sub_group_id;
+
+            IndexType ind_start = group_id * elems_for_sbg;
+            IndexType ind_end =
+                sycl::min(static_cast<IndexType>((group_id + 1) * elems_for_sbg), elem_count);
+
+            IndexType offset[radix_range_];
+
+            for (std::uint32_t i = 0; i < radix_range_; i++) {
+                offset[i] = part_prefix_hist_ptr[group_id * radix_range_ + i] +
+                            part_prefix_hist_ptr[n_total_sub_groups * radix_range_ + i];
+            }
+
+            for (IndexType i = ind_start + local_id; i < ind_end; i += local_size) {
+                radix_integer_t data_value = val_in_ptr[i];
+                radix_integer_t data_bits = ((inv_bits(data_value) >> bit_offset) & radix_range_1_);
+                IndexType pos_new = 0;
+                for (std::uint32_t j = 0; j < radix_range_; j++) {
+                    IndexType value = static_cast<IndexType>(data_bits == j);
+                    IndexType boundary = exclusive_scan(sbg, value, plus<IndexType>());
+                    pos_new |= value * (offset[j] + boundary);
+                    IndexType partial_offset = reduce(sbg, value, plus<IndexType>());
+                    offset[j] = offset[j] + partial_offset;
+                }
+                val_out_ptr[pos_new] = data_value;
+                ind_out_ptr[pos_new] = ind_in_ptr[i];
+            }
+        });
+    });
+
+    return event;
+}
+
+template <typename Float, typename IndexType>
+radix_sort_indices_inplace<Float, IndexType>::radix_sort_indices_inplace(sycl::queue& queue,
+                                                                         std::int64_t elem_count)
+        : queue_(queue),
+          elem_count_(0) {
+    init(queue, elem_count);
+}
+
+template <typename Float, typename IndexType>
+radix_sort_indices_inplace<Float, IndexType>::~radix_sort_indices_inplace() {
+    sort_event_.wait_and_throw();
+}
+
+template <typename Float, typename IndexType>
+void radix_sort_indices_inplace<Float, IndexType>::init(sycl::queue& queue,
+                                                        std::int64_t elem_count) {
+    std::uint32_t uint_elem_count = de::integral_cast<std::uint32_t>(elem_count);
+    if (uint_elem_count && elem_count_ != uint_elem_count) {
```

Comment:
Shouldn't we throw exception if conversion is impossible?

Reply:
Added additional asserts to check for ranges, this is primitive and checks with exceptions should be done in high level func